src/SignalR/src/Common/ReflectionHelper.cs

```diff
@@ -26,7 +26,7 @@ public static bool IsStreamingType(Type type)
             // TODO #2594 - add Streams here, to make sending files easy
             while (type != null)
             {
-                if (type.GetGenericTypeDefinition() == typeof(ChannelReader<>))
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(ChannelReader<>))
```

Comment:
Did you add any tests where `type` is a generic type that doesn't derive from `ChannelReader<>`?

Also nit: the `if (!type.IsGenericType)` block above can be removed now that you check this in the loop.

Reply:
Is this comment still valid:

```
             // IMPORTANT !!
             // All valid types must be generic
            // because HubConnectionContext gets the generic argument and uses it to determine the expected item type of the stream
            // The long-term solution is making a (streaming type => expected item type) method.
```

IMO the stream type doesn't have to be generic. If a developer has my `MyStringyChannelReader`, and it inherits from `ChannelReader<string>`, the type parameter should come from the base `ChannelReader<T>` type argument.

In fact getting it from the parent type could be an error. There could be a type like `WierdChannelReader<TSomethingElse> : ChannelReader<string>`. If a method got or returned `WierdChannelReader<int>` then the type would incorrectly be int instead of string.

Reply:
I'll remove the comment, but pretty sure our check is correct. Are you just showing an example based on the comment and not saying there is a bug in the code?

Reply:
There is a bug. In DefaultHubDispatcher.cs:

```cs
var itemType = methodExecutor.MethodParameters[i].ParameterType.GetGenericArguments()[0];
args[i] = connection.StreamTracker.AddStream(placeholder.StreamId, itemType);
```

IMO change IsStreamingType return the generic argument (like IsChannelType does), save it to the HubMethodDescriptor and use in the dispatcher. Will fix the bug and will improve performance slightly because the reflection is cached once.

Unit test for a non generic `ChannelReader<T>` implementation and a channel reader where the type argument isn't used by the reader.

Reply:
Oh I see what you mean. I'm not sure we want to support any type except types we explicitly say are ok. So for first pass, only `ChannelReader<T>`. Which means we can remove the loop in `IsStreamingType` and make the check simpler.

Reply:
Every streaming Hub method will need to take `ChannelReader<T>` directly. We won't support `WeirdChannelReader<int> : ChannelReader<string>` as an argument. So if we change `IsStreamingType` to check the top level type instead of the loop we have currently then there wont be any issue, and `WeirdChannelReader` will fail to bind as expected.

Reply:
True true. Don't change `IsStreamingType` though, it is used to process incoming parameters where inheritance does matter.

Reply:
`IsStreamingType` is only used when looking at hub method parameters and since that has to be `ChannelReader<T>` to be true then there is no reason to recurse.

Sadly if someone wrote:
```
public class ChannelThing : System.Threading.Channels.ChannelReader<string>
{
    public override bool TryRead(out string item)
    {
        // ...
    }

    public override ValueTask<bool> WaitToReadAsync(CancellationToken cancellationToken = default)
    {
        // ...
    }
}

public MyHub : Hub
{
    public void Send(ChannelThing thing)
    {
    }
}
```

The method will be called because the serializer ignores { "StreamId": 1 } when creating `ChannelThing`. I'll look at if I can force it to fail when we know the client is sending a stream and the hub method doesn't have a valid stream type in the parameters.

Reply:
Filed https://github.com/aspnet/AspNetCore/issues/4401 for this